[Core] add `[p]set api list` and `[p]set api remove`

Vuks69 opened the pull request:
### Type

- [ ] Bugfix
- [ ] Enhancement
- [X] New feature

### Description of the changes
Added new subcommands: `[p]set api list` and `[p]set api remove`.
`[p]set api list` causes the bot to send a message containing all currently set API services, without exposing tokens. The message is formatted similarly to `[p]repo list`.

`[p]set api remove` removes the provided services from config, including their associated tokens.

Should be useful for managing your API keys along with `[p]set api` and especially when removing api keys, since there were no easy-to-use tools for that.

Vuks69 committed:
add get_shared_api_keys

Vuks69 committed:
add set listapi command

Vuks69 committed:
fix typo

Vuks69 committed:
refactor `[p]set listapi` into `[p]set api list`

Vuks69 committed:
remove unnecessary check

Vuks69 committed:
fix decorators

Vuks69 committed:
corrected wording

Vuks69 committed:
add remove_shared_api_services

Vuks69 committed:
add `set api remove`

Vuks69 committed:
further typo sniping

Vuks69 committed:
bugsniping

Vuks69 committed:
minor typo

Vuks69 committed:
aaaaaaaaa

Drapersniper commented on redbot/core/core_commands.py:
```diff
@@ -2059,7 +2060,7 @@ async def custominfo(self, ctx: commands.Context, *, text: str = None):
+    @_set.group(invoke_without_command=True)
     @checks.is_owner()
     async def api(self, ctx: commands.Context, service: str, *, tokens: TokenConverter):
         """Set various external API tokens.
```
There is no mention subcomments of `[p]help set`, I would reccomend this command is broken up into 3 subcommands.

Jackenmen replied:
I think it would be better to just mention in `[p]set api`'s docstring that it's also for listing and deleting (in whatever way that would be worded) if there's an issue with clarity, so that all the commands related to API tokens are still grouped together.

Vuks69 replied:
I've updated the `[p]set api` docstring to reflect the new subcommands.

Drapersniper commented on redbot/core/core_commands.py:
```diff
@@ -2077,6 +2078,32 @@ async def api(self, ctx: commands.Context, service: str, *, tokens: TokenConvert
+            await ctx.send(_("No API services have been set yet."))
+            return
+
+        sorted_services = sorted(services, key=lambda r: str.lower(r))
```
```suggestion
        sorted_services = sorted(services, key=str.lower)
```

Drapersniper commented on redbot/core/core_commands.py:
```diff
@@ -2077,6 +2078,32 @@ async def api(self, ctx: commands.Context, service: str, *, tokens: TokenConvert
+        sorted_services = sorted(services, key=lambda r: str.lower(r))
+        joined = _("Set API services:\n")
+        for key in sorted_services:
+            joined += "{}\n".format(key)
+        for page in pagify(joined, ["\n"], shorten_by=16):
+            await ctx.send(box(page.lstrip(" "), lang="diff"))
```
```suggestion
            joined += "+ {}\n".format(key)
        for page in pagify(joined, ["\n"], shorten_by=16):
            await ctx.send(box(page.lstrip(" "), lang="diff"))
```
So that it applies some syntax highlighting
```diff
Set API services:
+ test
+ test2
```

Drapersniper commented on redbot/core/core_commands.py:
```diff
@@ -2077,6 +2078,32 @@ async def api(self, ctx: commands.Context, service: str, *, tokens: TokenConvert
+    @api.command(name="remove")
+    async def api_remove(self, ctx: commands.Context, *services: str):
+        """Remove the given services with all their keys and tokens."""
+        services = list(services)
```
```suggestion
```
This is not needed

Drapersniper reviewed (dismissed):
Pausing the review as @Vuks69  said he needs to look into something else.

Vuks69 committed:
Apply suggestions from code review

Co-authored-by: Draper <27962761+Drapersniper@users.noreply.github.com>

Vuks69 committed:
update api docstring to reflect new subcommands

Vuks69 commented:
From my side the PR is now ready for review.

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -2077,6 +2078,31 @@ async def api(self, ctx: commands.Context, service: str, *, tokens: TokenConvert
+    async def api_list(self, ctx: commands.Context):
+        """Show all external API services that have been set.
+
+        This command does not expose keys or tokens associated with services,
```
Hmm, it would (IMO) be useful to also know the key names that were set for the API services (i.e. `twitch: client_id, client_secret`), but I'm not gonna request it here - it's up to you whether you want to include it.

Vuks69 replied:
I will definitely look into that, but at a later date.

Jackenmen replied:
Based on a conversation in Discord, this will be made as part of this PR due to needed changes in the API methods that are added by this PR.

Jackenmen commented on redbot/core/bot.py:
```diff
@@ -1051,6 +1058,25 @@ async def remove_shared_api_tokens(self, service_name: str, *token_names: str):
+        Parameters
+        ----------
+        service_names: str
+            the services to remove
```
```suggestion
            The services to remove.
```

Jackenmen commented on redbot/core/core_commands.py:
```diff
@@ -2077,6 +2078,31 @@ async def api(self, ctx: commands.Context, service: str, *, tokens: TokenConvert
+            return
+
+        sorted_services = sorted(services, key=str.lower)
+        joined = _("Set API services:\n")
```
Plural support:
```suggestion
        joined = _("Set API services:\n") if len(services) > 1 else _("Set API service:\n")
```

Jackenmen reviewed (changes requested):
Aside from review comments, you also need to add the new methods (`Red.remove_shared_api_services`, `Red.get_shared_api_services`) at the end of `docs/framework_apikeys.rst` file. It should be easy to figure out from other methods there how you need to add the new ones but do let me know if you run into any issues with that or anything else.

Vuks69 committed:
Apply suggestions from code review

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

Vuks69 committed:
update framework_apikeys.rst with new methods

Jackenmen commented on redbot/core/bot.py:
```diff
@@ -1051,6 +1058,25 @@ async def remove_shared_api_tokens(self, service_name: str, *token_names: str):
             for name in token_names:
                 group.pop(name, None)
 
+    async def remove_shared_api_services(self, service_names: Iterable[str]):
```
I've just noticed that unlike `remove_shared_api_tokens()`, this method is using an iterable as a parameter. I think it would be better for consistency if this did it similarly to the aforementioned method:
```suggestion
    async def remove_shared_api_services(self, *service_names: str):
```

// I'm just posting the code suggestion for the method signature; the method usage and its docstring also need to be updated and I haven't made the code suggestion for that.

Sorry for missing this earlier :(

Vuks69 committed:
Update redbot/core/bot.py

Co-authored-by: jack1142 <6032823+jack1142@users.noreply.github.com>

Vuks69 committed:
rewrite get_shared_api_services into a special case of
get_shared_api_tokens

Vuks69 committed:
rewrite api_list to include keys and tokens in response

Vuks69 committed:
don't show secrets

Vuks69 committed:
better api_remove response

Vuks69 committed:
black

Vuks69 committed:
remove nonexistent method

Vuks69 committed:
remove unnecessary import

Vuks69 committed:
fix wording

Jackenmen committed:
Use `humanize_list()` rather than `str.join()`

This is done for consistency within Red
and it makes the list have the last element joined
with `and` (or its equivalent in chosen locale)

Jackenmen committed:
Use `.format()` after translation is applied

If `.format()` is used before `_()`, the search for translation is done
with the string that already has the dynamic text added,
which results in no matches.

Jackenmen committed:
Improve error message

Updated message to be more specific
(used phrases: "None of the services" and "had any keys set")
and a bit more nice to the user (judging word "anyway" removed)

Jackenmen committed:
Improve readability of `[p]set api list`

It's a lot clearer when the sublists are indented,
especially on mobile where code blocks aren't colored at all.

New look:
https://user-images.githubusercontent.com/6032823/94614944-3bbd7c80-02a7-11eb-89e4-64bdf06c650b.png

Jackenmen committed:
Improve docstrings and type hints

- added overloads to help out developers a bit more
  * this wasn't necessary, but development tools work better
    with this information
- fixed type hint to use Union as now
  its return type depends on whether `service_name` is passed
- updated docstrings to contain information about the added behavior

Jackenmen committed:
Add plural support

Jackenmen reviewed (approved):
Looks good to me after doing few changes.

@Vuks69 I said that I'm gonna take care of the rest here, but since you might want to look at what I've done, I put a full list at the end of this and I'm gonna wait with merging this until tomorrow.

---

**Improve readability of `[p]set api list`**

It's a lot clearer when the sublists are indented,
especially on mobile where code blocks aren't colored at all.

New look:
https://user-images.githubusercontent.com/6032823/94614944-3bbd7c80-02a7-11eb-89e4-64bdf06c650b.png

---

**Improve error message**

Updated message to be more specific
(used phrases: "None of the services" and "had any keys set")
and a bit more nice to the user (judging word "anyway" removed)

---

**Add plural support**

---

**Use `.format()` after translation is applied**

If `.format()` is used before `_()`, the search for translation is done
with the string that already has the dynamic text added,
which results in no matches.

---

**Use `humanize_list()` rather than `str.join()`**

This is done for consistency within Red
and it makes the list have the last element joined
with `and` (or its equivalent in chosen locale)

---

**Improve docstrings and type hints**

- added overloads to help out developers a bit more
  * this wasn't necessary, but development tools work better
    with this information
- fixed type hint to use Union as now
  its return type depends on whether `service_name` is passed
- updated docstrings to contain information about the added behavior
